### ISSUE TITLE
impl Trait that does not capture generic argument

### ISSUE THREAD

**Issue by jonhoo** [rust-lang/rust#65442]
I'd like to write the following code:
```rust
use std::future::Future;
struct Foo<A>(A);
impl<A> Foo<A> {
    fn bar<Q, R: Default>(&mut self, q: Q) -> impl Future<Output = R> {
        let _ = q;
        async move {
            R::default()
        }
    }
    
    fn baz(&mut self, x: &str) -> impl Future<Output = usize> {
        self.bar(x)
    }
}
```
In particular, I would like to have the `impl Trait` returned by `baz` _not_ be tied to the lifetime of its `&str` argument. Since `impl Trait` captures the lifetimes of all generic arguments (as per [RFC 1951](https://github.com/rust-lang/rfcs/blob/master/text/1951-expand-impl-trait.md)), I can't write the code this way though. So instead, I tried
```rust
#![feature(type_alias_impl_trait)]
use std::future::Future;
struct Foo<A>(A);

type BarFut<A, R> = impl Future<Output = R>;
impl<A> Foo<A> {
    fn bar<Q, R: Default>(&mut self, q: Q) -> BarFut<A, R> {
        let _ = q;
        async move {
            R::default()
        }
    }
    
    fn baz(&mut self, x: &str) -> impl Future<Output = usize> {
        self.bar(x)
    }
}
```
However, with this, I get the error:
```
error: type parameter `Q` is part of concrete type but not used in parameter list for the `impl Trait` type alias
```
This seems odd, since `Q` is (intentionally) not used in the `async` block. I can work around this by adding an `async fn` and calling that instead of using `async move`, but that seems like an odd hack:
```rust
async fn make_r_fut<R: Default>() -> R {
    R::default()
}
// ...
    fn bar<Q, R: Default>(&mut self, q: Q) -> BarFut<A, R> {
        let _ = q;
        make_r_fut()
    }
// ...
```

Is it intentional that the `async` block "captures" `Q` here, even though it never contains a `Q`?

**Comment by Mark-Simulacrum** [comment]
I suspect the answer is no -- cc @rust-lang/wg-async-await -- this might contribute to "not Send/Sync" errors if we're capturing what we shouldn't be

**Comment by nikomatsakis** [comment]
I think the answer is probably *no* but I'm not sure how hard this would be to fix.

**Comment by nikomatsakis** [comment]
Assigning to myself to investigate and document the cause of the issue, at least.

**Comment by Arnavion** [comment]
Related to / dupe of https://github.com/rust-lang/rust/issues/42940

**Comment by glittershark** [comment]
I hate bumping issues but... any movement on this? Just ran into it myself in a very similar context

**Comment by nikomatsakis** [comment]
No progress I know of

**Comment by hamza1311** [comment]
Note that this bug does not require async to trigger. [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7aaadf37f899a192e15b3c6dff23ba72) where this triggered using `type_alias_impl_trait`. I'm not sure if/how this is related #42940 (as mentioned in the comment above). Looking through the discussion there, I couldn't find any mentions of the error returned by the compiler.

**Comment by SOF3** [comment]
Reproduced with a more minimal example: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=319e66180308bede359b022cda8ed0bf

```rust
#![feature(type_alias_impl_trait)]

type Ty = impl Fn();

fn run<E>(_: E) -> Ty {
    || ()
}
```

```
   Compiling playground v0.0.1 (/playground)
error: type parameter `E` is part of concrete type but not used in parameter list for the `impl Trait` type alias
 --> src/lib.rs:6:5
  |
6 |     || ()
  |     ^^^^^

error: could not compile `playground` due to previous error```

**Comment by traviscross** [comment]
Here's the minimum `async` example of this problem:

```rust
#![feature(type_alias_impl_trait)]

use std::future::Future;

type Deferred = impl Future<Output = ()> + 'static;

fn future_out<F>(_: F) -> Deferred {
    async move { }
}
```

```
error: type parameter `F` is part of concrete type but not used in parameter list for the `impl Trait` type alias
 --> src/lib.rs:8:5
  |
8 |     async move { }
  |     ^^^^^^^^^^^^^^

error: could not compile `playground` (lib) due to previous error
```

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=32c1651fa2a10fd03251ee0df9aabb4a).

**Comment by traviscross** [comment]
Notably, there's discussion over in #107645 about the possibility of limiting TAIT to exclude defining uses where the TAIT appears in argument position in the function signature.  If that were to be adopted, one plausible workaround would look like this:

```rust
#![feature(type_alias_impl_trait)]

use std::future::Future;
use tokio::sync::oneshot::Sender;

type Deferred = impl Future<Output = ()> + 'static;

fn sends_future_kludge<F: Future<Output = ()> + 'static>(f: F) -> Deferred {
    async move { f.await }
}

fn sends_future(tx: Sender<Deferred>) {
    let _ = tx.send(sends_future_kludge(async move {}));
}
```

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=312eda7e8840f2c7e53e308f24130b6f).

However, that workaround would be blocked on this issue.

**Comment by Dylan-DPC** [comment]
Current output (of [minimised code](https://github.com/rust-lang/rust/issues/65442#issuecomment-1242717664))

```
error: unconstrained opaque type
 --> src/lib.rs:3:11
  |
3 | type Ty = impl Fn();
  |           ^^^^^^^^^
  |
  = note: `Ty` must be used in combination with a concrete type within the same crate

error[E0308]: mismatched types
 --> src/lib.rs:6:5
  |
3 | type Ty = impl Fn();
  |           --------- the expected opaque type
4 |
5 | fn run<E>(_: E) -> Ty {
  |                    -- expected `Ty` because of return type
6 |     || ()
  |     ^^^^^ expected opaque type, found closure
  |
  = note: expected opaque type `Ty`
                 found closure `{closure@src/lib.rs:6:5: 6:7}`
```

**Comment by WaffleLapkin** [comment]
The output is unchanged actually. It's just that since #128440 you need to add an attribute to define taits:
```rust
#![feature(type_alias_impl_trait)]

type Ty = impl Fn();

#[define_opaque(Ty)]
fn run<E>(_: E) -> Ty {
    || ()
}
```
```
error: type parameter `E` is part of concrete type but not used in parameter list for the `impl Trait` type alias
 --> src/lib.rs:6:20
  |
6 | fn run<E>(_: E) -> Ty {
  |                    ^^
```
[[play]](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=8b131ef1f97e46ee21c273a5546b68fa)

---

```rust
#![feature(type_alias_impl_trait)]

use std::future::Future;

type Deferred = impl Future<Output = ()> + 'static;

#[define_opaque(Deferred)]
fn future_identity<F>(_: F) -> Deferred {
    async move { }
}
```
```
error: type parameter `F` is part of concrete type but not used in parameter list for the `impl Trait` type alias
 --> src/lib.rs:8:32
  |
8 | fn future_identity<F>(_: F) -> Deferred {
  |                                ^^^^^^^^
```
[[play]](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=e9d9f58cf036b22e3fe31ea69a80b65f)
